Library/ViewModels/ReportProjectFormViewModel.swift

```diff
@@ -5,117 +5,108 @@ import ReactiveSwift
 
 public protocol ReportProjectFormViewModelInputs {
   func viewDidLoad()
-
-  /// Submits a report using the createFlagging mutation
-  func submitReport(
-    contentId: String,
-    kind: GraphAPI.FlaggingKind,
-    details: String,
-    clientMutationId: String?
-  )
 }
 
-public protocol ReportProjectFormViewModelOutputs {
-  /// Emits the currently logged in user's email
-  var userEmail: Signal<String, Never> { get }
-}
+public protocol ReportProjectFormViewModelOutputs {}
 
 public protocol ReportProjectFormViewModelType {
   var inputs: ReportProjectFormViewModelInputs { get }
   var outputs: ReportProjectFormViewModelOutputs { get }
 }
 
 public final class ReportProjectFormViewModel: ReportProjectFormViewModelType,
-  ReportProjectFormViewModelInputs,
-  ReportProjectFormViewModelOutputs, ObservableObject {
-  public var bannerMessage: PassthroughSubject<MessageBannerViewViewModel, Never> = .init()
-  public var detailsText: PassthroughSubject<String, Never> = .init()
-  public var projectID: PassthroughSubject<String, Never> = .init()
-  public var projectFlaggingKind: PassthroughSubject<GraphAPI.FlaggingKind, Never> = .init()
-  public var retrievedEmail: PassthroughSubject<String, Never> = .init()
-  public var saveButtonEnabled: AnyPublisher<Bool, Never>
-  public var saveTriggered: PassthroughSubject<Bool, Never> = .init()
-  public var submitSuccess: PassthroughSubject<Void, Never> = .init()
+  ReportProjectFormViewModelInputs, ReportProjectFormViewModelOutputs, ObservableObject {
+  @Published public var retrievedEmail: String? = nil
+  @Published public var saveButtonEnabled: Bool = false
+  @Published public var detailsText: String = ""
+  @Published public var saveTriggered: Bool = false
+  @Published public var bannerMessage: MessageBannerViewViewModel? = nil
 
-  private var cancellables = Set<AnyCancellable>()
+  @Published public var submitSuccess: Bool = false
 
-  public init() {
-    self.saveButtonEnabled = Publishers
-      .CombineLatest(self.projectFlaggingKind, self.detailsText)
-      .map { !$0.1.isEmpty }
-      .eraseToAnyPublisher()
+  private let viewDidLoadSubject = PassthroughSubject<Bool, Never>()
+
+  private var cancellables = Set<AnyCancellable>()
 
-    let userEmailEvent = self.viewDidLoadProperty.signal
-      .switchMap { _ in
+  public let projectID: String
+  public let projectURL: String
+  public let projectFlaggingKind: GraphAPI.FlaggingKind
+
+  public init(projectID: String,
+              projectURL: String,
+              projectFlaggingKind: GraphAPI.FlaggingKind) {
+    self.projectID = projectID
+    self.projectURL = projectURL
+    self.projectFlaggingKind = projectFlaggingKind
+
+    /// Only enable the save button if the user has entered detail text
+    self.$detailsText
+      .map { !$0.isEmpty }
+      .assign(to: &$saveButtonEnabled)
```

Comment:
This is an example of a nice, very brief binding between Combine + SwiftUI. Elegant!

Reply:
Oh this is pretty cool how `.assign(to` manages the life cycle of the subscription automatically so we don't have to keep track of an AnyCancellable

Reply:
Yeah, I really dig this pattern